src/resample2d.js

```diff
@@ -0,0 +1,147 @@
+          ih = 0;
+        } else if (ih > srcHeight - 1) {
+          ih = srcHeight - 1;
+        }
+        for (let ow = 0; ow < dstWidth; ow++) {
+          let iw = (ow + 0.5) / scaleWidth - 0.5;
+          if (iw < 0) {
+            iw = 0;
+          } else if (iw > srcWidth - 1) {
+            iw = srcWidth - 1;
+          }
+          const outputValue = src.getValueByLocation(
+              [ob, Math.ceil(ih - 0.5), Math.ceil(iw - 0.5), oc]);
+          dst.setValueByLocation([ob, oh, ow, oc], outputValue);
+        }
+      }
+    }
+  }
+  return dst;
+}
+
+/**
+ * Resample the tensor values from the source to the destination spatial dimensions according to
+ * the scaling factors using Bilinear interpolation.
+ * Refer to https://en.wikipedia.org/wiki/Bilinear_interpolation
+ * @param {Tensor} src
+ * @param {Tensor} dst
+ * @param {Number} dstHeight
+ * @param {Number} dstWidth
+ * @param {Number} scaleHeight
+ * @param {Number} scaleWidth
+ * @return {Tensor}
+ */
+function linear(src, dst, dstHeight, dstWidth, scaleHeight, scaleWidth) {
+  const [srcBatches, srcHeight, srcWidth, srcChannels] = src.shape;
+  for (let ob = 0; ob < srcBatches; ob++) {
+    for (let oc = 0; oc < srcChannels; oc++) {
+      for (let oh = 0; oh < dstHeight; oh++) {
+        let ih = (oh + 0.5) / scaleHeight - 0.5;
+        if (ih < 0) {
+          ih = 0;
+        } else if (ih > srcHeight - 1) {
+          ih = srcHeight - 1;
+        }
+        const ihLower = Math.floor(ih);
+        const ihUpper = Math.ceil(ih);
+        const u = ih - ihLower;
+        for (let ow = 0; ow < dstWidth; ow++) {
+          let iw = (ow + 0.5) / scaleWidth - 0.5;
+          if (iw < 0) {
+            iw = 0;
+          } else if (iw > srcWidth - 1) {
+            iw = srcWidth - 1;
+          }
+          const iwLower = Math.floor(iw);
+          const iwUpper = Math.ceil(iw);
+          const v = iw - iwLower;
+          const outputValue =
+              (1 - u) * (1 - v) * src.getValueByLocation([ob, ihLower, iwLower, oc]) +
+              (1 - u) * v * src.getValueByLocation([ob, ihLower, iwUpper, oc]) +
+              u * (1 - v) * src.getValueByLocation([ob, ihUpper, iwLower, oc]) +
+              u * v * src.getValueByLocation([ob, ihUpper, iwUpper, oc]);
+          dst.setValueByLocation([ob, oh, ow, oc], outputValue);
+        }
+      }
+    }
+  }
+  return dst;
+}
+
+const interpolationFunctions = {
+  'nearest-neighbor': nearestNeighbor,
+  'linear': linear,
+};
+
+/**
+ * Resample the tensor values from the source to the destination spatial dimensions according to
+ * the scaling factors.
+ * @param {Tensor} input
+ * @param {MLResample2dOptions} options
+ * @return {Tensor}
+ */
+export function resample2d(
+    input,
+    {
+      mode = 'nearest-neighbor',
+      scales = [1.0, 1.0],
+      sizes,
+      axes = [2, 3],
+    } = {}) {
+  if (axes[0] === 0) {
+    // hwnc -> nhwc
+    input = transpose(input, {permutation: [2, 0, 1, 3]});
```

Comment:
It would be more robust to verify that `axes[0] == 0 && axes[1] == 1`, rather than just `axes[0]`.

I'd also add an `else` for `axes[0] == 1 && axes[1] == 2` for clarity, even if it is just a comment that says "NHWC -> NHWC Do nothing". That way it's apparent to readers that you didn't leave it out.

Reply:
Btw, I had this clever idea that you could just say a single line...

`input = transpose(permutation: [(axes[0] + 2) & 3, (axes[1] + 2) & 3, axes[0], axes[1]]);`

...instead of explicit switch case statements for each transpose, and it works nicely and generically for the first transpose (it assigns the last 2 dimensions H and W to read from whatever axes were given, and it assigns the batch and channel to the other two axes). All the standard axes work:

`axes = [0,1] -> permute:[2, 3, 0, 1]`
`axes = [1,2] -> permute:[3, 0, 1, 2]`
`axes = [2,3] -> permute:[0, 1, 2, 3]`

And it works for the *mirrors* of the standard ones:

`axes = [1,0] -> permute:[3, 2, 1, 0]`
`axes = [2,1] -> permute:[0, 3, 2, 1]`
`axes = [3,2] -> permute:[1, 0, 3, 2]`

And even works for HNCW (which isn't valid per the spec anyway):

`axes = [0,3] -> permute:[2, 1, 0, 3]`

But it's not clear how to *untranspose* that (the 2nd transpose). So, I'm not proposing that idea...

Reply:
> input = transpose(permutation: [(axes[0] + 2) & 3, (axes[1] + 2) & 3, axes[0], axes[1]]);'

Thanks, this code is fairly concise. 
For more easily understand, I modified previous switch case statements following your first comment.